server/src/main/java/org/opensearch/gateway/GatewayAllocator.java

```diff
@@ -335,10 +396,54 @@ protected AsyncShardFetch.FetchResult<TransportNodesListShardStoreMetadata.NodeS
             }
             return shardStores;
         }
+    }
+
+    class InternalReplicaBatchShardAllocator extends ReplicaShardBatchAllocator {
```

Comment:
Can you implement `hasInitiatedFetching` function in this class or should I pick this whole Internal class in my PR?